app/controllers/file_manager_controller.rb

```diff
@@ -0,0 +1,234 @@
+        send_file absolute_path
+      else
+        @file = File.read(absolute_path)
+        @parent = './'
+        render :file, formats: :html
+      end
+    end
+  end
+
+  def upload
+    upload_file(params[:path].nil? ? "" : params[:path])
+  end
+
+  def delete
+    absolute_path = check_path_exist(params[:path])
+    if check_instructor(absolute_path)
+      if File.directory?(absolute_path)
+        FileUtils.rm_rf(absolute_path)
+      else
+        FileUtils.rm(absolute_path)
+      end
+    end
+    flash[:success] = "Successfully deleted"
+  end
+
+  def rename
+    absolute_path = check_path_exist(params[:relative_path])
+    if check_instructor(absolute_path)
+      if absolute_path.equal?(BASE_DIRECTORY)
+        flash[:error] = "Unable to rename courses in the root directory."
+      else
+        dir_name = File.dirname(params[:relative_path])
+
+        if params[:new_name].empty?
+          raise ArgumentError, "New name not provided,
+        new name cannot be blank"
+        end
+
+        unless params[:new_name].match(/^[a-zA-Z0-9_-]+(\.[a-zA-Z0-9_-]+)?$/)
+          raise ArgumentError, "Invalid characters. Only letters,
+        numbers, underscores, and hyphens are allowed."
+        end
+
+        new_path = safe_expand_path("#{dir_name}/#{params[:new_name]}")
+        parent = new_path.split[0..-2].join('/')
+
+        raise ArgumentError, "A file with that name already exists" if File.exist?(new_path)
+
+        FileUtils.mkdir_p(parent)
+        FileUtils.mv(absolute_path, new_path)
+        flash[:success] = "Successfully renamed file to #{params[:new_name]}"
+      end
+    end
+  rescue ArgumentError => e
+    flash[:error] = e.message
+  end
+
+  def download_tar
+    path = params[:path]&.split("/")&.drop(2)&.join("/")
+    path = CGI.unescape(path)
+    absolute_path = check_path_exist(path).to_s
+    return unless check_instructor(absolute_path)
+
+    if File.directory?(absolute_path)
+      tar_stream = StringIO.new("")
+      Gem::Package::TarWriter.new(tar_stream) do |tar|
+        Dir[File.join(absolute_path, '**', '**')].each do |file|
+          mode = File.stat(file).mode
+          relative_path = file.sub(%r{^#{Regexp.escape(absolute_path)}/?}, '')
+          if File.directory?(file)
+            tar.mkdir relative_path, mode
+          else
+            tar.add_file relative_path, mode do |tar_file|
+              File.open(file, "rb") { |f| tar_file.write f.read }
+            end
+          end
+        end
+      end
+      tar_stream.rewind
+      tar_stream.close
+      send_data tar_stream.string.force_encoding("binary"),
+                filename: "file_manager.tar",
+                type: "application/x-tar",
+                disposition: "attachment"
+    else
+      send_file(absolute_path,
+                filename: File.basename(absolute_path),
+                disposition: 'attachment')
+    end
+    flash[:success] = "Download successful"
```

Comment:
If you're making these calls through AJAX, the user won't see the success messages until they load another page. Consider displaying the messages in some other way (e.g. AJAX `success` method)

Reply:
I just removed this because there is a success message logged to console in the AJAX call anyways.